core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -608,6 +608,38 @@ object Chunk {
    * Returns a chunk from a number of values.
    */
   def apply[A](a: A, as: A*): NonEmptyChunk[A] =
+    as match {
+      case Nil =>
+        Singleton(a)
+      case wa: scala.collection.mutable.WrappedArray[A] =>
+        val len: Int        = wa.size + 1
+        val in: Array[A]    = wa.array.asInstanceOf[Array[A]]
+        val ct: ClassTag[A] = wa.elemTag.asInstanceOf[ClassTag[A]]
+
+        if (len <= 16) {
+
+          val dest: Array[A] = Array.ofDim[A](len)(ct)
```

Comment:
Why copy this? You can wrap it into a chunk directly, right?

Reply:
It is faster to copy it directly with the missing element `a`, even in the creation. It is faster also when using it after.

Reply:
Isn’t it faster we because we are capturing the existing class tag versus creating a new one? Is there a variant where we reuse the class tag but just create a new Concat with the first element and the existing array?

Reply:
strangely it cost a lot more, for `Chunk(1, 2, 3, 4)`: 
* 20 ns if you recopy
* 60+ ns if you `Concat(Singleton(a), Arr(element, tag))`

I had both methods on a benchmark : 
* new is `Concat`
* current is full recopy
* old is the one on master

the mix of those two methods at N=16 is coming from this benchmark.
```scala
[info] Benchmark                                  Mode  Cnt    Score     Error  Units
[info] ChunkCreationBenchmarks.arrayCreate01      avgt    5    4.936 ±   0.052  ns/op
[info] ChunkCreationBenchmarks.arrayCreate02      avgt    5    4.604 ±   0.145  ns/op
[info] ChunkCreationBenchmarks.arrayCreate04      avgt    5    5.742 ±   0.082  ns/op
[info] ChunkCreationBenchmarks.arrayCreate08      avgt    5    8.498 ±   0.105  ns/op
[info] ChunkCreationBenchmarks.arrayCreate16      avgt    5   13.976 ±   0.219  ns/op
[info] ChunkCreationBenchmarks.arrayCreate32      avgt    5   24.785 ±   0.053  ns/op
[info] ChunkCreationBenchmarks.chunkCreate01      avgt    5   45.557 ±   0.758  ns/op
[info] ChunkCreationBenchmarks.chunkCreate01_new  avgt    5   44.548 ±   2.560  ns/op
[info] ChunkCreationBenchmarks.chunkCreate01_old  avgt    5   46.464 ±   2.556  ns/op
[info] ChunkCreationBenchmarks.chunkCreate02      avgt    5   14.258 ±   0.123  ns/op
[info] ChunkCreationBenchmarks.chunkCreate02_new  avgt    5   58.595 ±   1.976  ns/op
[info] ChunkCreationBenchmarks.chunkCreate02_old  avgt    5   62.040 ±   4.300  ns/op
[info] ChunkCreationBenchmarks.chunkCreate04      avgt    5   17.231 ±   0.202  ns/op
[info] ChunkCreationBenchmarks.chunkCreate04_new  avgt    5   63.196 ±  39.850  ns/op
[info] ChunkCreationBenchmarks.chunkCreate04_old  avgt    5   66.586 ±   0.316  ns/op
[info] ChunkCreationBenchmarks.chunkCreate08      avgt    5   30.694 ±   1.041  ns/op
[info] ChunkCreationBenchmarks.chunkCreate08_new  avgt    5   57.755 ±   5.651  ns/op
[info] ChunkCreationBenchmarks.chunkCreate08_old  avgt    5   75.993 ±   4.792  ns/op
[info] ChunkCreationBenchmarks.chunkCreate16      avgt    5   61.813 ±   3.506  ns/op
[info] ChunkCreationBenchmarks.chunkCreate16_new  avgt    5   62.322 ±   4.343  ns/op
[info] ChunkCreationBenchmarks.chunkCreate16_old  avgt    5  106.836 ±   1.873  ns/op
[info] ChunkCreationBenchmarks.chunkCreate32      avgt    5  180.660 ± 187.046  ns/op
[info] ChunkCreationBenchmarks.chunkCreate32_new  avgt    5  116.431 ±  98.966  ns/op
[info] ChunkCreationBenchmarks.chunkCreate32_old  avgt    5  196.811 ± 183.724  ns/op
[info] ChunkCreationBenchmarks.vectorCreate01     avgt    5   78.172 ±  14.909  ns/op
[info] ChunkCreationBenchmarks.vectorCreate02     avgt    5   82.857 ±  27.224  ns/op
[info] ChunkCreationBenchmarks.vectorCreate04     avgt    5   98.103 ±   3.921  ns/op
[info] ChunkCreationBenchmarks.vectorCreate08     avgt    5  132.327 ±  12.716  ns/op
[info] ChunkCreationBenchmarks.vectorCreate16     avgt    5  205.143 ±   8.434  ns/op
[info] ChunkCreationBenchmarks.vectorCreate32     avgt    5  335.735 ±   8.287  ns/op
[success] Total time: 3033 s (50:33), completed 6 Apr 2020, 01:30:48

```

Reply:
The variant with Concat is when N>16 :
https://github.com/zio/zio/pull/3297/files#diff-290e96e6c64027a32c36d751a35d65deR631

Reply:
Sounds good!